ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,241 @@
+        
+        // Convert PHP string to UTF-16
+        UChar *ustr = NULL;
+        int32_t ustr_len = 0;
+        UErrorCode status = U_ZERO_ERROR;
+        
+        intl_convert_utf8_to_utf16(&ustr, &ustr_len, ZSTR_VAL(str_val), ZSTR_LEN(str_val), &status);
+        zend_string_release(str_val);
+
+        if (U_FAILURE(status)) {
+            // We can't use goto cleanup because items and itemLengths are incompletely allocated
+            for (uint32_t j = 0; j < i; j++) {
+                efree((void *)items[j]);
+            }
+            efree(items);
+            efree(itemLengths);
+            intl_error_set(NULL, status, "Failed to convert string to UTF-16", 0);
+            RETURN_FALSE;
+        }
+        
+        items[i] = ustr;
+        itemLengths[i] = ustr_len;
+        i++;
+    } ZEND_HASH_FOREACH_END();
+
+    UErrorCode status = U_ZERO_ERROR;
+    int32_t resultLength;
+    UChar *result = NULL;
+
+    resultLength = ulistfmt_format(LISTFORMATTER_OBJECT(obj), items, itemLengths, count, NULL, 0, &status);
+
+    if (U_FAILURE(status) && status != U_BUFFER_OVERFLOW_ERROR) {
+        intl_error_set(NULL, status, "Failed to format list", 0);
+        RETVAL_FALSE;
+        goto cleanup;
+    }
+
+    // Allocate buffer and try again
+    status = U_ZERO_ERROR;
+    result = (UChar *)safe_emalloc(resultLength + 1, sizeof(UChar), 0);
+    ulistfmt_format(LISTFORMATTER_OBJECT(obj), items, itemLengths, count, result, resultLength, &status);
+
+    if (U_FAILURE(status)) {
+        if (result) {
+            efree(result);
+        }
+        intl_error_set(NULL, status, "Failed to format list", 0);
+        RETVAL_FALSE;
+        goto cleanup;
+    }
+
+    // Convert result back to UTF-8
+    zend_string *ret = intl_convert_utf16_to_utf8(result, resultLength, &status);
+    efree(result);
+    
+    if (!ret) {
+        intl_error_set(NULL, status, "Failed to convert result to UTF-8", 0);
+        RETVAL_FALSE;
+    } else {
+        RETVAL_NEW_STR(ret);
+    }
+
+    cleanup:
+        for (i = 0; i < count; i++) {
+            efree((void *)items[i]);
+        }
+        efree(items);
+        efree(itemLengths);
+}
+
+PHP_METHOD(IntlListFormatter, getErrorCode)
+{
+    ZEND_PARSE_PARAMETERS_NONE();
+
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+
+    UErrorCode status = intl_error_get_code(LISTFORMATTER_ERROR_P(obj));
+
+    RETURN_LONG(status);
+}
+
+PHP_METHOD(IntlListFormatter, getErrorMessage)
+{
+     ZEND_PARSE_PARAMETERS_NONE();
+
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+
+    zend_string *message = intl_error_get_message(LISTFORMATTER_ERROR_P(obj));
+    RETURN_STR(message);
+}
+
+void listformatter_register_class(void)
+{
+    zend_class_entry *class_entry = register_class_IntlListFormatter();
+    class_entry->create_object = listformatter_create_object;
+    
+    memcpy(&listformatter_handlers, zend_get_std_object_handlers(), sizeof(zend_object_handlers));
```

Comment:
Since clone_obj is not set, cloning is allowed, but that won't actually work would it?

Reply:
I've disabled cloning and added a test for it. Unfortunately, we don't have a ulistfmt_clone function like we do for the other formatters.